File: in_Java/0417 Pacific Atlantic Water Flow.java
class Solution {
    int[][] dir = {{0,1},{1,0},{0,-1},{-1,0}};
    public List<List<Integer>> pacificAtlantic(int[][] matrix) {
        List<List<Integer>> res = new ArrayList<>();
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) return res;
        int row = matrix.length, col = matrix[0].length;
        int[][] m1 = new int[row][col];
        int[][] m2 = new int[row][col];
        /*
        for(int i = 0; i < row; i++){
            dfs(matrix,m1,i,0);
            dfs(matrix,m2,i,col-1);
        }
        for(int i = 0; i < col; i++){
            dfs(matrix,m1,0,i);
            dfs(matrix,m2,row-1,i);
        }
        */
        
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(i == 0 || j == 0){
                    dfs(matrix,m1,i,j);
                }
            }
        }
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(i == row-1 || j == col-1){
                    dfs(matrix,m2,i,j);
                }
            }
        }
        
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(m1[i][j] == 1 && m2[i][j] == 1){
                    List<Integer> t = new ArrayList<>();
                    t.add(i);t.add(j);
                    res.add(t);
                }
            }
        }
        return res;
    }
    private void dfs(int[][] m, int[][] v, int i, int j){
        v[i][j] = 1;
        for(int[] d : dir){
            int x = i + d[0];
            int y = j + d[1];
            if(x < 0 || y < 0 || x >= m.length || y >= m[0].length || m[i][j] > m[x][y] || v[x][y] == 1) continue;
            dfs(m,v,x,y);
        }
    }
}

File: in_Python/0417 Pacific Atlantic Water Flow.py
class Solution(object):
    def pacificAtlantic(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[List[int]]
        """
        res = []
        if matrix == None or len(matrix) == 0 or len(matrix[0]) == 0:
            return res
        pac_mat = []
        atl_mat = []
        row, col = len(matrix), len(matrix[0])
        directions = ((0,1),(0,-1),(1,0),(-1,0))
        for r in range(row):
            t = []
            tt = []
            for c in range(col):
                t.append(0)
                tt.append(0)
            pac_mat.append(tt)
            atl_mat.append(t)
        for r in range(row):
            for c in range(col):
                if r == 0 or c == 0:
                    pac_mat[r][c] = 1
                if r == row - 1 or c == col - 1:
                    atl_mat[r][c] = 1
        for i in range(row):
            for j in range(col):
                q = collections.deque()
                q.append((i,j))
                while q:
                    c = q.popleft()
                    curi, curj = c[0], c[1]
                    for di in directions:
                        nexti = curi + di[0]
                        nextj = curj + di[1]
                        if 0 <= nexti < row and 0 <= nextj < col:
                            if matrix[nexti][nextj] >= matrix[curi][curj] and pac_mat[nexti][nextj] == 0 and pac_mat[curi][curj] == 1:
                                q.append((nexti,nextj))
                                pac_mat[nexti][nextj] = 1
        for i in range(row-1, -1, -1):
            for j in range(col-1, -1, -1):
                q = collections.deque()
                q.append((i,j))
                while q:
                    c = q.popleft()
                    curi, curj = c[0], c[1]
                    for di in directions:
                        nexti = curi + di[0]
                        nextj = curj + di[1]
                        if 0 <= nexti < row and 0 <= nextj < col:
                            if matrix[nexti][nextj] >= matrix[curi][curj] and atl_mat[nexti][nextj] == 0 and atl_mat[curi][curj] == 1:
                                q.append((nexti,nextj))
                                atl_mat[nexti][nextj] = 1
        for i in range(row):
            for j in range(col):
                if pac_mat[i][j] == 1 and atl_mat[i][j] == 1:
                    res.append([i,j])
        return res
                                    
                    
                        
                
                